code/modules/recycling/disposal/pipe.dm

```diff
@@ -0,0 +1,364 @@
+
+/obj/structure/disposalpipe/proc/transfer_to_dir(obj/structure/disposalholder/H, nextdir)
+	H.setDir(nextdir)
+	var/turf/T = H.nextloc()
+	var/obj/structure/disposalpipe/P = H.findpipe(T)
+
+	if(P)
+		// find other holder in next loc, if inactive merge it with current
+		var/obj/structure/disposalholder/H2 = locate() in P
+		if(H2 && !H2.active)
+			H.merge(H2)
+
+		H.loc = P
+		return P
+	else			// if wasn't a pipe, then they're now in our turf
+		H.loc = get_turf(src)
+		return null
+
+// update the icon_state to reflect hidden status
+/obj/structure/disposalpipe/proc/update()
+	var/turf/T = src.loc
+	hide(T.intact && !isspaceturf(T))	// space never hides pipes
+
+// hide called by levelupdate if turf intact status changes
+// change visibility status and force update of icon
+/obj/structure/disposalpipe/hide(var/intact)
+	invisibility = intact ? INVISIBILITY_MAXIMUM: 0	// hide if floor is intact
+
+// expel the held objects into a turf
+// called when there is a break in the pipe
+/obj/structure/disposalpipe/proc/expel(obj/structure/disposalholder/H, turf/T, direction)
+	var/turf/target
+	var/eject_range = 5
+	var/turf/open/floor/floorturf
+
+	if(isfloorturf(T)) //intact floor, pop the tile
+		floorturf = T
+		if(floorturf.floor_tile)
+			new floorturf.floor_tile(T)
+		floorturf.make_plating()
+
+	if(direction)		// direction is specified
+		if(isspaceturf(T)) // if ended in space, then range is unlimited
+			target = get_edge_target_turf(T, direction)
+		else						// otherwise limit to 10 tiles
+			target = get_ranged_target_turf(T, direction, 10)
+
+		eject_range = 10
+
+	else if(floorturf)
+		target = get_offset_target_turf(T, rand(5)-rand(5), rand(5)-rand(5))
+
+	playsound(src, 'sound/machines/hiss.ogg', 50, 0, 0)
+	for(var/atom/movable/AM in H)
+		AM.forceMove(src.loc)
+		AM.pipe_eject(direction)
+		if(target)
+			AM.throw_at(target, eject_range, 1)
+	H.vent_gas(T)
+	qdel(H)
+
+
+// pipe affected by explosion
+/obj/structure/disposalpipe/contents_explosion(severity, target)
+	var/obj/structure/disposalholder/H = locate() in src
+	if(H)
+		H.contents_explosion(severity, target)
+
+
+/obj/structure/disposalpipe/run_obj_armor(damage_amount, damage_type, damage_flag = 0, attack_dir)
+	if(damage_flag == "melee" && damage_amount < 10)
+		return 0
+	. = ..()
+
+
+//attack by item
+//weldingtool: unfasten and convert to obj/disposalconstruct
+/obj/structure/disposalpipe/attackby(obj/item/I, mob/user, params)
+	var/turf/T = src.loc
+	if(T.intact)
+		return		// prevent interaction with T-scanner revealed pipes
+	add_fingerprint(user)
+	if(istype(I, /obj/item/weldingtool))
+		if(!can_be_deconstructed(user))
+			return
+
+		var/obj/item/weldingtool/W = I
+		if(W.remove_fuel(0, user))
+			playsound(src, I.usesound, 50, 1)
+			to_chat(user, "<span class='notice'>You start slicing [src]...</span>")
+			// check if anything changed over 2 seconds
+			if(do_after(user, 30*I.toolspeed, target = src))
+				if(!src || !W.isOn())
```

Comment:
!src => !QDELETED(src)

Reply:
Actually, `do_after` checks for `QDELETED` already.